### ISSUE TITLE
Don't publish one chado record under multiple content types.

### ISSUE THREAD

**Issue by dsenalik** [tripal/tripal#1984]
### Task Description

Right now, as with Tripal 3, if you have a "generic" content type and a "specific" content type both using the same chado table, a record may be published twice.
For example, if you manually enter a "Genome Project" entity, and then publish the more generic "Project" content type, it will be published as a generic "Project" entity.
We don't have this problem with content types derived from the `feature` table, as in this case there is no generic "Feature" content type.

I am thinking that publish could get a list of all content types from a particular base table, get all those terms, and then it would have to check for a `type_id` column in the base table, or if that doesn't exist, look in the property table (as for example Project uses).

### Branch Name

tv4g1-issue1984-only-publish-once

**Comment by Ferrisx4** [comment]
A discussion in Slack about this topic came across a few possible solutions to this. The examples use the analysis type and its subtypes, typically Genome Assembly and Genome Annotation, but also apply to other "generic" types that have "specific" subtypes such as Project and Genome Project mentioned above.

1. Create a generic type for "Analysis", assigned to the existing analysis bundle. This way all content from the analysis table belongs to a specific bundle. A proposed term to use to identify these "generic" analyses could be `Analysis (operation:2945)`. We'd need an update hook to address sites that do not have this value stored in `analysisprop` and add the Chado Type Reference field to the Analysis bundle. 
2. When publishing generic analyses, exclude analyses that have a specified subtype in the analysisprop table. This may be a little less straightforward because the type_id in the analysisprop table refers to the (and matches) the term itself. Because of this, we'd have to maintain a list of terms (`value`) or `term_id`s that designate the entity as a subtype and aren't just some other property e.g. `is_obsolete`.
3. Just for completeness of record from the conversation, an unlikely approach would be to, when publishing Analysis content, also publish all content of all subtypes and publish them under their respective types. This would then prompt the user when publishing Analysis content that *"This may also publish Genome Annotation and Genome Assembly content"*. This is challenging as @dsenalik pointed out because the table is determined from the bundle, and not the other way around.

**Comment by Ferrisx4** [comment]
A 4th approach not mentioned would be to simply not publish content that is already published, however this would create a situation where the order of publishing matters. From the example above, a site would have to publish all specific analyses (Genome Assembly, Genome Annotation, etc.) first before publishing generic Analysis content.

**Comment by dsenalik** [comment]
Here is a potential workaround I am testing that a site admin could apply to a site.
1. Add the term to the analysis bundle. This can be done easily by "+Check for new fields" and adding this one
![issue1984 1](https://github.com/user-attachments/assets/e87c3e87-ea97-476d-9a2a-b4e21e6283aa)

2. All migrated "plain" analyses need to have this property added, so some SQL is needed. Output will be the count of the number of records with properties added.
```
DROP FUNCTION IF EXISTS update_types;
CREATE FUNCTION update_types() RETURNS integer as $$
DECLARE
  _loop_id bigint;
  _count bigint;
BEGIN
  _count := 0;
  FOR _loop_id IN
    SELECT analysis_id FROM analysis
    WHERE analysis_id NOT IN (
      SELECT A.analysis_id FROM analysis A
        LEFT JOIN analysisprop P on A.analysis_id=P.analysis_id WHERE P.type_id IN (
          SELECT cvterm_id FROM cvterm T WHERE name IN ('Genome annotation', 'genome assembly', 'Analysis')
          AND cv_id=(SELECT cv_id FROM cv WHERE name='EDAM')
        )
    )
  LOOP
    INSERT INTO analysisprop (analysis_id, type_id, value)
      VALUES (_loop_id,
              (SELECT T.cvterm_id FROM cvterm T LEFT JOIN cv CV ON T.cv_id=CV.cv_id
                   WHERE T.name = 'Analysis' AND CV.name = 'EDAM'),
              'Analysis'
             );
    _count := _count + 1;
  END LOOP;
  RETURN _count;
END;
$$ LANGUAGE plpgsql;


SELECT update_types();
```

3. Similarly for project
![issue1984 2](https://github.com/user-attachments/assets/6b930df7-4391-44c6-a4b0-d08d750260a8)
4. SQL for project
```
DROP FUNCTION IF EXISTS update_types;
CREATE FUNCTION update_types() RETURNS integer as $$
DECLARE
  _loop_id bigint;
  _count bigint;
BEGIN
  _count := 0;
  FOR _loop_id IN
    SELECT project_id FROM project 
    WHERE project_id NOT IN (
      SELECT A.project_id FROM project A
        LEFT JOIN projectprop P on A.project_id=P.project_id WHERE P.type_id IN (
          SELECT cvterm_id FROM cvterm T WHERE name IN ('Genome Project', 'Project')
        )
    )
  LOOP
    INSERT INTO projectprop (project_id, type_id, value)
      VALUES (_loop_id,
              (SELECT T.cvterm_id FROM cvterm T LEFT JOIN cv CV ON T.cv_id=CV.cv_id
                   WHERE T.name = 'Project' AND CV.name = 'ncit'),
              'Project'
             );
    _count := _count + 1;
  END LOOP;
  RETURN _count;
END;
$$ LANGUAGE plpgsql;


SELECT update_types();
```

**Comment by dsenalik** [comment]
I have tested both of these on my migrated site, and there is now no overlap when publishing!
These are the only two tables in core tripal with this issue, stock and feature have multiple bundle types, but no "generic" one.

**Comment by Ferrisx4** [comment]
After applying the same fixes (with a few tweaks to the function, below), I'm now able to publish 30 analyses and not 448 (the rest are subtypes). Awesome! 

Slightly altered script for analysis content:
 - I added `chado.` to the table references because where I ran the script doesn't pick up the name of the chado schema automatically
 - Added `Gene functional annotation` for the i5k site. 
```sql
DROP FUNCTION IF EXISTS update_types;
CREATE FUNCTION update_types() RETURNS integer as $$
DECLARE
  _loop_id bigint;
  _count bigint;
BEGIN
  _count := 0;
  FOR _loop_id IN
    SELECT analysis_id FROM chado.analysis
    WHERE analysis_id NOT IN (
      SELECT A.analysis_id FROM chado.analysis A
        LEFT JOIN chado.analysisprop P on A.analysis_id=P.analysis_id WHERE P.type_id IN (
          SELECT cvterm_id FROM chado.cvterm T WHERE name IN ('Genome annotation', 'genome assembly', 'Analysis', 'Gene functional annotation')
          AND cv_id=(SELECT cv_id FROM chado.cv WHERE name='EDAM')
        )
    )
  LOOP
    INSERT INTO chado.analysisprop (analysis_id, type_id, value)
      VALUES (_loop_id,
              (SELECT T.cvterm_id FROM chado.cvterm T LEFT JOIN chado.cv CV ON T.cv_id=CV.cv_id
                   WHERE T.name = 'Analysis' AND CV.name = 'EDAM'),
              'Analysis'
             );
    _count := _count + 1;
  END LOOP;
  RETURN _count;
END;
$$ LANGUAGE plpgsql;


SELECT update_types();
```
---
So the workaround works, which is exciting. Interesting that the analysis type reference field isn't added by default. I'm on board with this approach. I'd even say we try to incorporate it into Tripal (with @laceysanderson's blessing).

For the edit I made to the function to add Gene functional annotation, that works, but I'm wondering if we wanted to automatically pick up those bundle names if it would be possible, for other sites that may have added further subtypes to any of the default types.

**Comment by dsenalik** [comment]
I agree that we should add the analysis + project type reference, although there still will be need of the SQL when migrating a tripal 3 site, so it's not something an update hook could do unless your site was already migrated and not yet published, an unlikely combination.

For picking up bundle names, that gets complicated and can't be done just in SQL. Maybe we could make a drush command ...
